src/miner.cpp

```diff
@@ -483,52 +456,83 @@ void static BitcoinMiner(CWallet *pwallet)
+                // Find valid nonce
+                while (true)
                 {
-                    if (UintToArith256(hash) <= hashTarget)
-                    {
-                        // Found a solution
-                        pblock->nNonce = nNonce;
-                        assert(hash == pblock->GetHash());
+                    // H(I||V||...
+                    crypto_generichash_blake2b_state curr_state;
+                    curr_state = state;
+                    crypto_generichash_blake2b_update(&curr_state,
+                                                      pblock->nNonce.begin(),
+                                                      pblock->nNonce.size());
+
+                    // (x_1, x_2, ...) = A(I, V, n, k)
+                    LogPrint("pow", "Running Equihash solver with nNonce = %s\n",
+                             pblock->nNonce.ToString());
+                    std::set<std::vector<unsigned int>> solns = eh.BasicSolve(curr_state);
+                    LogPrint("pow", "Solutions: %d\n", solns.size());
+
+                    // Write the solution to the hash and compute the result.
+                    for (auto soln : solns) {
+                        pblock->nSolution = soln;
+
+                        if (UintToArith256(pblock->GetHash()) > hashTarget) {
+                            continue;
+                        }
 
+                        // Found a solution
                         SetThreadPriority(THREAD_PRIORITY_NORMAL);
-                        LogPrintf("BitcoinMiner:\n");
-                        LogPrintf("proof-of-work found  \n  hash: %s  \ntarget: %s\n", hash.GetHex(), hashTarget.GetHex());
+                        LogPrintf("ZcashMiner:\n");
+                        LogPrintf("proof-of-work found  \n  hash: %s  \ntarget: %s\n", pblock->GetHash().GetHex(), hashTarget.GetHex());
                         ProcessBlockFound(pblock, *pwallet, reservekey);
                         SetThreadPriority(THREAD_PRIORITY_LOWEST);
 
                         // In regression test mode, stop mining after a block is found.
                         if (chainparams.MineBlocksOnDemand())
                             throw boost::thread_interrupted();
 
-                        break;
+                        goto updateblock;
                     }
+                    pblock->nNonce = ArithToUint256(UintToArith256(pblock->nNonce) + 1);
+                    if ((UintToArith256(pblock->nNonce) & 0xfff) == 0)
+                        break;
                 }
 
+updateblock:
                 // Check for stop or if block needs to be rebuilt
                 boost::this_thread::interruption_point();
                 // Regtest mode doesn't require peers
                 if (vNodes.empty() && chainparams.MiningRequiresPeers())
                     break;
-                if (nNonce >= 0xffff0000)
+                if (UintToArith256(pblock->nNonce) >= 0xffff0000)
```

Comment:
Yes it's a wall-clock estimation; the `0xfff` check in particular was chosen so that these checks above happened every few seconds. Obviously that won't be the case initially for this PoW, so I'll make it check after every solution, at least initially.